monitor/monitor.go

```diff
@@ -0,0 +1,260 @@
+package monitor
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"sync"
+	"time"
+
+	log "github.com/Sirupsen/logrus"
+
+	"github.com/9corp/9volt/alerter"
+	"github.com/9corp/9volt/config"
+	"github.com/9corp/9volt/dal"
+	"github.com/9corp/9volt/util"
+)
+
+const (
+	START int = iota
+	STOP
+
+	GOROUTINE_ID_LENGTH = 8
+)
+
+type IMonitor interface {
+	Run() error
+	Stop()
+	Identify() string
+}
+
+type Monitor struct {
+	Config             *config.Config
+	Identifier         string
+	runningMonitorLock *sync.Mutex
+	runningMonitors    map[string]IMonitor
+	MessageChannel     chan *alerter.Message
+	SupportedMonitors  map[string]func(*RootMonitorConfig) IMonitor // monitor name : NewXMonitor
+}
+
+type RootMonitorConfig struct {
+	GID            string // goroutine id
+	Name           string
+	Config         *MonitorConfig
+	MessageChannel chan *alerter.Message
+	StopChannel    chan bool
+	Ticker         *time.Ticker
+}
+
+// TODO: This should probably be split up between each individual check type
+type MonitorConfig struct {
+	// Generic attributes that fit more than one monitor type
+	Type        string // 'tcp', 'http', 'ssh', 'exec', 'icmp', 'dns'
+	Description string // optional
+	Host        string // required for all checks except 'exec'
+	Interval    util.CustomDuration
+	Timeout     util.CustomDuration
+	Port        int    // works for all checks except 'icmp' and 'exec'
+	Expect      string // works for 'tcp', 'ssh', 'http', 'exec' checks except 'icmp'
+	Enabled     bool
+	Tags        []string
+
+	// HTTP specific attributes
+	HTTPURL         string `json:"URL"`
+	HTTPMethod      string `json:"Method"`
+	HTTPSSL         bool   `json:"SSL"`
+	HTTPStatusCode  int    `json:"StatusCode"`
+	HTTPRequestBody string `json:"RequestBody"` // Only used if 'Method' is 'GET'
+
+	// Exec specific attributes
+	ExecCommand    string `json:"Command"`
+	ExecReturnCode int    `json:"ReturnCode"`
+
+	// Alerting related configuration
+	WarningThreshold  int      // how many times a check must fail before a warning alert is emitted
```

Comment:
Maybe this later turns into a monitor context specific element. Ie. A `WarningThreshold` could be `300ms` for a specific type of `http` check. Something to think about.